symengine/serialize-cereal.h

```diff
@@ -0,0 +1,223 @@
+#pragma once
+#include <symengine/basic.h>
+#include <symengine/integer.h>
+#include <symengine/symbol.h>
+#include <symengine/visitor.h>
+#include <symengine/utilities/stream_fmt.h>
+
+#include <cereal/archives/binary.hpp>
+#include <cereal/cereal.hpp>
+#include <cereal/types/polymorphic.hpp>
+#include <cereal/types/string.hpp>
+#include <cereal/details/helpers.hpp>
+#include <cereal/types/map.hpp>
+#include <cereal/types/unordered_map.hpp>
+
+namespace SymEngine
+{
+
+template <class Archive>
+inline void save_basic(Archive &ar, const Basic &b)
+{
+    const auto t_code = b.get_type_code();
+    throw std::runtime_error(StreamFmt() << __FILE__ << ":" << __LINE__ << ": " << __PRETTY_FUNCTION__
+                                 << " not supported: " << type_code_name(t_code) << " (" << t_code << ")"
+#if !defined(NDEBUG)
+                             << ", " << b.__str__()
+#endif
+        );
+}
+template <class Archive>
+inline void save_basic(Archive &ar, const Symbol &b)
+{
+    ar(b.__str__());
+}
+template <class Archive>
+inline void save_basic(Archive &ar, const Pow &b)
+{
+    ar(b.get_base());
+    ar(b.get_exp());
+}
+template <class Archive>
+inline void save_basic(Archive &ar, const Add &b)
+{
+    ar(b.get_coef());
+    ar(b.get_dict());
+}
+template <class Archive>
+inline void save_basic(Archive &ar, const Integer &b)
+{
+    ar(b.__str__());
+}
+
+// template <class Archive>
+// inline void save_basic(Archive &ar, const Rational &b)
+// {
+//     ar(b.get_num(), b.get_den());
+// }
+
+#define SYMENGINE_ENUM(type, Class)                     \
+    template <class Archive>                            \
+    inline void save_basic(Archive &ar, const Class &b) \
```

Comment:
You can use `save_basic(Archive &ar, const OneArgFunction &b)` and get rid of defining these multiple times.

Reply:
Ah, of course, will fix this.

Reply:
So this really reduced the number of save_basic implementations. For `load_basic` I haven't yet come up with anything as elegant. I have the gut feeling it should be possible to do something leaner, any ideas?

Reply:
`load_basic(Archive &ar, RCP<const OneArgFunction> &b)` ?

Reply:
OneArgFunction is abstract, so I don't think that's possible?

Reply:
I mean I can dispatch on it, but then I'm at loss:
```c++
template <class Archive>
RCP<const Basic> load_basic(Archive &ar, RCP<const OneArgFunction> &)
{
    RCP<const Basic> arg;
    ar(arg);
    return basic_rcp_from_type_code_and_args(type_code, arg);
}
```
I guess I could change the signature of all `load_basic` dispatch functions to take `type_code` as an arguemnt, but then we still need the hypothetical `basic_rcp_from_type_code_and_args`?

Reply:
See https://github.com/symengine/symengine/pull/1704/commits/f7ada3705aa89797a764ec84b20a582b1e599ff7